state/mongodb/mongodb.go

```diff
@@ -192,20 +205,70 @@ func (m *MongoDB) setInternal(ctx context.Context, req *state.SetRequest) error
 	if req.ETag != nil {
 		filter[etag] = *req.ETag
 	} else if req.Options.Concurrency == state.FirstWrite {
-		filter[etag] = uuid.NewString()
+		uuid, err := uuid.NewRandom()
+		if err != nil {
+			return err
+		}
+		filter[etag] = uuid.String()
+	}
+
+	reqTTL, err := stateutils.ParseTTL(req.Metadata)
+	if err != nil {
+		return fmt.Errorf("failed to parse TTL: %w", err)
 	}
 
-	update := bson.M{"$set": bson.M{id: req.Key, value: v, etag: uuid.NewString()}}
-	_, err := m.collection.UpdateOne(ctx, filter, update, options.Update().SetUpsert(true))
+	etagV, err := uuid.NewRandom()
+	if err != nil {
+		return err
+	}
 
-	return err
+	update := mongo.Pipeline{
+		{{"$set", bson.D{
+			{id, req.Key},
+			{value, v},
+			{etag, etagV.String()},
+		}}},
+	}
+
+	if reqTTL != nil {
+		update = append(update, primitive.D{{"$addFields", bson.D{
+			{daprttl, bson.D{
+				{"$dateAdd",
+					bson.D{{"startDate", "$$NOW"}, {"unit", "second"}, {"amount", *reqTTL}},
+				},
+			}},
+		}}})
+	} else {
+		update = append(update, primitive.D{{"$addFields", bson.D{
+			{daprttl, nil},
+		}}})
+	}
+
+	_, err = m.collection.UpdateOne(ctx, filter, update, options.Update().SetUpsert(true))
+	if err != nil {
+		return fmt.Errorf("error in updating document: %s", err)
+	}
+
+	return nil
 }
 
 // Get retrieves state from MongoDB with a key.
 func (m *MongoDB) Get(ctx context.Context, req *state.GetRequest) (*state.GetResponse, error) {
+	// Since MongoDB doesn't delete the document immediately when the TTL value
+	// is reached, we need to filter out the documents with TTL value less than
+	// the current time.
+	filter := bson.D{
```

Comment:
Random thought: "NoSQL" may be nice but looking at this really makes me miss a nice `SELECT * FROM table WHERE ttl < CURRENT_TIME` 🙃

Reply:
:upside_down_face: :upside_down_face: :upside_down_face: